atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java

```diff
@@ -119,33 +124,180 @@ public void shouldOnlyReturnHostsMatchingPredicate() {
+
+        assertThat(cassandraClientPool.getRandomUntriedHost(ImmutableSet.of(host1, host2)),
+                is(Optional.empty()));
+    }
+
+    @Test
+    public void shouldRedirectToPreferredHostsIfAvailable() {
+        InetSocketAddress host1 = new InetSocketAddress(HOSTNAME_1, DEFAULT_PORT);
+        InetSocketAddress host2 = new InetSocketAddress(HOSTNAME_2, DEFAULT_PORT);
+        CassandraClientPool cassandraClientPool =
+                MockCassandraClientPoolUtils.clientPoolWithServersInCurrentPool(ImmutableSet.of(host1, host2));
+
+        for (int i = 0; i < FUZZING_NUM_TRIALS; i++) {
+            CassandraClientPoolingContainer container
+                    = cassandraClientPool.getRedirectTarget(ImmutableSet.of(), ImmutableSet.of(host1));
+            assertThat(container.getHost(), equalTo(host1));
+        }
+    }
+
+    @Test
+    public void shouldRedirectToOtherPreferredHostAfterTryingOne() {
+        InetSocketAddress host1 = new InetSocketAddress(HOSTNAME_1, DEFAULT_PORT);
+        InetSocketAddress host2 = new InetSocketAddress(HOSTNAME_2, DEFAULT_PORT);
+        InetSocketAddress host3 = new InetSocketAddress(HOSTNAME_3, DEFAULT_PORT);
+        CassandraClientPool cassandraClientPool =
+                MockCassandraClientPoolUtils.clientPoolWithServersInCurrentPool(ImmutableSet.of(host1, host2, host3));
+
+        for (int i = 0; i < FUZZING_NUM_TRIALS; i++) {
+            CassandraClientPoolingContainer container
+                    = cassandraClientPool.getRedirectTarget(ImmutableSet.of(host1), ImmutableSet.of(host1, host2));
+            assertThat(container.getHost(), equalTo(host2));
+        }
+    }
+
+    @Test
+    public void shouldRedirectToNonPreferredHostsAfterTryingAllPreferred() {
+        InetSocketAddress host1 = new InetSocketAddress(HOSTNAME_1, DEFAULT_PORT);
+        InetSocketAddress host2 = new InetSocketAddress(HOSTNAME_2, DEFAULT_PORT);
+        CassandraClientPool cassandraClientPool =
+                MockCassandraClientPoolUtils.clientPoolWithServersInCurrentPool(ImmutableSet.of(host1, host2));
+
+        for (int i = 0; i < FUZZING_NUM_TRIALS; i++) {
+            CassandraClientPoolingContainer container
+                    = cassandraClientPool.getRedirectTarget(ImmutableSet.of(host1), ImmutableSet.of(host1));
+            assertThat(container.getHost(), equalTo(host2));
+        }
+    }
+
+    @Test
+    public void shouldBeAbleToRedirectAfterTryingAllHosts() {
+        InetSocketAddress host1 = new InetSocketAddress(HOSTNAME_1, DEFAULT_PORT);
+        InetSocketAddress host2 = new InetSocketAddress(HOSTNAME_2, DEFAULT_PORT);
+        CassandraClientPool cassandraClientPool =
+                MockCassandraClientPoolUtils.clientPoolWithServersInCurrentPool(ImmutableSet.of(host1, host2));
+
+        cassandraClientPool.getRedirectTarget(ImmutableSet.of(host1, host2), ImmutableSet.of(host1));
```

Comment:
What guarantees that it choose `host1` here? I thought it would try a random host?